File: Dynamic_Programming/1223. Dice Roll Simulation.cpp
/*
1223. Dice Roll Simulation
*/

// T:O(6*n*k*6*k) SC:O(6*n*k)
class Solution {
    using LL = long long;
    LL M = 1e9+7;
public:
    int dieSimulator(int n, vector<int>& rollMax) {
        auto dp = vector<vector<vector<int>>>(n, vector<vector<int>>(6,vector<int>(16,0)));
        for(int j = 0; j < 6; j++){
            dp[0][j][1] = 1;
        }

        for(int i = 1; i < n; i++){
            for(int j = 0; j < 6; j++){
                for(int k =1; k <= rollMax[j]; k++){
                    if(k > 1){
                        dp[i][j][k] = dp[i-1][j][k-1];
                    }
                    else {
                        for(int jj = 0; jj < 6; jj++){
                            if(j == jj) continue;//如果只投出一次不會有 dp[i][相同數字][1~max]的繼承過來
                            for(int kk = 1; kk <= rollMax[jj]; kk++){
                                dp[i][j][k] += dp[i-1][jj][kk];
                                dp[i][j][k]%=M;
                            }
                        }
                    }
                }
            }
        }
        int res = 0;
        for(int j = 0; j < 6; j++){
            for(int k = 1; k <= rollMax[j]; k++){
                res += dp[n-1][j][k];
                res%=M;
            }
        }

        return res;

    }
};

/*
此題給你一串數組說明1-6不能投出超過這麼多連續的次數 有幾種情況是合法的
每種情況有幾種可能都是取決於上一輪狀態  
dp[i] 第幾輪 [j] 投到哪個數字 [k] 此數字出現幾次
*/

File: Dynamic_Programming/1223. Dice Roll Simulation.py
# T:O(n * 6 * 6 * k) SC:O(6 * k)
class Solution:
    def dieSimulator(self, n: int, rollMax: List[int]) -> int:
        '''
        每個數字出現的連續次數不能超過各自的limit 那會有多少種可能性
        n = 5000 總共會丟這麼多次, rollMax = 15,上一個是誰 總共出現了幾次
        '''
        MOD = 10 ** 9 + 7
        r_max = max(rollMax)
        f = [[0] * (r_max + 1) for _ in range(6)]
        for i in range(6):
            f[i][1] = 1
        for i in range(1, n):
            new_f = [[0] * (r_max + 1) for _ in range(6)]
            for pre in range(6):
                for cur in range(6):
                    if cur != pre:
                        # 當更換骰子面時，連續次數重設為 1
                        new_f[cur][1] = (new_f[cur][1] + sum(f[pre][cnt] for cnt in range(1, min(rollMax[pre], i + 1) + 1))) % MOD
                    else:
                        # 當保持骰子面不變時，連續次數增加 1，但必須不超過該面限制
                        for cnt in range(1, min(rollMax[pre], i + 1)):
                            new_f[cur][cnt + 1] = (new_f[cur][cnt + 1] + f[cur][cnt]) % MOD
            f = new_f
        return sum(sum(row) for row in f) % MOD

        '''
        @cache
        def dfs(i:int, pre:int, cnt:int) ->int:
            if i == n:
                return 1
            ans = 0
            for j in range(1, 7):
                if j != pre:
                    ans = (ans + dfs(i+1, j, 1)) % MOD
                elif j == pre and cnt + 1 <= rollMax[j-1]:
                    ans = (ans + dfs(i+1, j, cnt + 1)) % MOD
            return ans
        return dfs(0, -1, 0)
        '''
